sematic/caching/caching.py

```diff
@@ -0,0 +1,106 @@
+# Standard Library
+import hashlib
+import json
+import logging
+from typing import Any, Callable, Optional, Union
+
+# Sematic
+from sematic.abstract_future import AbstractFuture
+from sematic.types.serialization import value_to_json_encodable
+
+CacheNamespaceCallable = Callable[[AbstractFuture], str]
+CacheNamespace = Optional[Union[str, CacheNamespaceCallable]]
+
+logger = logging.getLogger(__name__)
+
+
+def get_future_cache_key(
+    cache_namespace: CacheNamespace, future: AbstractFuture
+) -> str:
+    """
+    Generates a cache key that can be used to uniquely identify the output value of a
+    deterministic function.
+
+    Parameters
+    ----------
+    cache_namespace: CacheNamespace
+        A string or a `Callable` which returns a string, which is used as the cache key
+        namespace.
+    future: AbstractFuture
+        The future for which to calculate the cache key.
+
+    Returns
+    -------
+    A cache key string that can be used to uniquely identify the output value of a
+    deterministic function.
+    """
+    # this may raise, so do it early
+    cache_namespace = _resolve_namespace(cache_namespace=cache_namespace, future=future)
+
+    func_fqpn = future.calculator.get_func_fqpn()  # type: ignore # noqa: ignore
+    output_type_repr = repr(future.calculator.output_type)
+    input_args_hash = _get_input_args_hash(future)
+
+    hash_base = f"{cache_namespace}|{func_fqpn}|{output_type_repr}|{input_args_hash}"
+    cache_key = _hash(hash_base)
```

Comment:
Instead of the final key being:
```
cache_key = _hash(concat(namespace, other_pieces))
```
can we make it:
```
cache_key = concat(namespace, _hash(other_pieces))
```

Rationale: there is a use case where you see some previously completed execution and say "oh, I would like to use the caching results from that!" How do you know which namespace you use to have a shot at cache hits from it? You'll need to know what namespace it used. You won't be able to do that if the namespace got blurred into a hash itself.

Reply:
In general, +1 to the most human-readable cache key possible, in only for debugging purposes.